### DISCUSSION TITLE
How to reset state properly?

### DISCUSSION THREAD

**Discussion by ahmadxgani** [pmndrs/zustand#1869] [Q&A] [answered]
I confused why the dataForm not reset correctly when the resetDataForm invoked
`dataFormSlice.js`
```js
export const createDataFormSlice = (set) => ({
  dataForm: {},
  updatePrimitive: (key, value) => set(state => ({
    dataForm: {
      ...state.dataForm,
      [key]: value
    }
  })),
  resetDataForm: () => set({ dataForm: {} })
})
```

the resetDataForm invoked once the component unmounted
`Mutate.js`
```js
useEffect(() => {
  if (!variants) {
    updateBasicData('skus', [{ ...JSON.parse(JSON.stringify(sku)), code: self.crypto.randomUUID() }])
  }

  return () => {
    resetDataForm()
  }
}, [])

useEffect(() => {
  console.log(dataForm)
}, [dataForm])
```

and when i navigate to another route (which is perform cleanup function and invoke resetDataForm()
but the value of dataForm still exists:
![image](https://github.com/pmndrs/zustand/assets/73988956/b0dbd208-300f-4fec-a8e9-b1d7bf67005d)

i put `console.log(dataForm)` in `Index.js` component

**Comment by dbritto-dev** [comment]
@ahmadxgani if you are working with react router you need to use `useLocation` in order to listen changes on a route some routes are not unmounted just updated so I would use this:

```ts
const location = useLocation()

useEffect(() => {
  resetDataForm()
}, [location])
```

**Reply by dbritto-dev** [reply]
@ahmadxgani keep me in the loop

**Reply by ahmadxgani** [reply]
thanks for your answer, and i apologize for the late response

i try your suggestion, but seems not work, skus object still exists
i assume that it's because the dataForm filled after `updateBasicData`, and `useEffect` didn't realize that cuz the deps is empty `[]`
and i try to add deps but still not work, here's the updated `useEffect`:
```
useEffect(() => {
  if (!variants) {
    updateBasicData('skus', [{ ...JSON.parse(JSON.stringify(sku)), code: self.crypto.randomUUID() }])
  }

  return () => {
    console.log(dataForm)
    if (Object.keys(dataForm).length) resetDataForm()
  }
}, [dataForm.skus])
```

![image](https://github.com/pmndrs/zustand/assets/73988956/607fb3c2-a809-44e6-a4fc-bc259ba025f5)

from the image above, i realize that the `Index.js:65` log first instead of `console.log` inside cleanup function (`Mutate.js:53`), and i'm confused how to organize the render order so that cleanup function executed first. any idea @dbritto-dev ?

**Reply by dbritto-dev** [reply]
@ahmadxgani would you mind sharing us a basic example of your implementation on codesandbox? I would be happy to troubleshoot it and help you

**Reply by ahmadxgani** [reply]
here's the [codesandbox](https://codesandbox.io/s/zustand-example-forked-2522hn?file=/src/components/FirstPage.tsx), the first or initial render of second page contains `skus` object, which shouldn't exist

**Reply by dbritto-dev** [reply]
@ahmadxgani here you go -> https://codesandbox.io/s/zustand-example-forked-sqlqyp?file=/src/components/FirstPage.tsx

BTW, I updated your code to use `useLocation()` that's the right way to sync something when route changed

**Reply by ahmadxgani** [reply]
thanks for your quick response @dbritto-dev!
![image](https://github.com/pmndrs/zustand/assets/73988956/d7f79826-0852-4dc7-9fd2-0ef7cdb18c46)
however, the issue isn't resolved, in the first initial render of second page, the data not resetted :(

[updated codesandbox](https://codesandbox.io/s/zustand-example-forked-fmny4h?file=/src/components/FirstPage.tsx)

**Reply by dbritto-dev** [reply, marked as answer]
@ahmadxgani my bad I forgot to save my changes ... here you go -> https://codesandbox.io/s/zustand-example-forked-sqlqyp?file=/src/components/FirstPage.tsx

BTW, the right way to track changes on your store is by using `subscribe()`

**Reply by ahmadxgani** [reply]
![image](https://github.com/pmndrs/zustand/assets/73988956/8d4b9ecf-2080-41a4-95d2-aedb27f50075)
is it not possible to reset the state so that the initial render/first render is empty? in the example you gave, it resets after the 2nd render.

Btw, thank you for being bearable with me

**Reply by dbritto-dev** [reply]
> ![image](https://github.com/pmndrs/zustand/assets/73988956/8d4b9ecf-2080-41a4-95d2-aedb27f50075)
> is it not possible to reset the state so that the initial render/first render is empty? in the example you gave, it resets after the 2nd render.
> 
> Btw, thank you for being bearable with me
> 

Is not possible since we're resetting our data when route changes and the event occurs after component is mounted.

And the only way to do it is by using `useLocation()` hope that helps

**Reply by ahmadxgani** [reply]
thank you for your help @dbritto-dev! now my confusion has been answered.
